demo/Demo.hpp

```diff
@@ -0,0 +1,130 @@
+     */
+    bool IsValid() const;
+
+    /**
+     * @brief Computes and renders frame
+     */
+    void RunFrame();
+
+    /**
+     * @brief Makes physical and render instance of the line and returns it
+     * @param[in] particle physical data
+     * @param[in] start first point of the line
+     * @param[in] end last point of the line
+     * @return created object reference
+     */
+    Object& MakeLine(Particle particle, glm::vec3 start, glm::vec3 end);
+
+    /**
+     * @brief Makes physical and render instance of the plane and returns it
+     * @param[in] particle physical data
+     * @param[in] normal normal of the vector
+     * @return created object references
+     */
+    Object& MakePlane(Particle particle, glm::dvec3 normal);
+
+    /**
+     * @brief Makes physical and render instance of the sphere and returns it
+     * @param[in] particle physical data
+     * @param[in] radius radius of the sphere
+     * @return created object reference
+     */
+    Object& MakeSphere(Particle particle, double radius);
+
+    /**
+     * @brief Makes physical and render instance of the bo and returns it
+     * @param[in] particle physical data
+     * @param[in] i orthogonal basis vector of the box base
+     * @param[in] j orthogonal basis vector of the box base
+     * @param[in] k orthogonal basis vector of the box base
+     * @return created object refernce
+     */
+    Object& MakeBox(Particle particle, glm::vec3 i, glm::vec3 j, glm::vec3 k);
+
+    /**
+     * @brief Removes object from the demo
+     * @param[in] object reference to the object to be deleted
+     */
+    void Remove(Object& object);
+
+    //! Maximum number of particles in the demo
+    uint32_t const maxParticles = 200;
+
+    /**
+     * @brief Represents an instance of the render and physical objects
+     */
+    struct Object
+    {
+        Object(RigidBody* body, std::unique_ptr<render::primitive::Primitive>&& renderShape);
+
+        //! Physical body
+        RigidBody* body;
+
+        //! Render data
+        std::unique_ptr<render::primitive::Primitive> shape;
+    };
+
+private:
+    std::list<Object> m_objects;
+    render::Renderer& m_pRenderer;
+    ParticleWorld m_particleWorld;
+    Particles m_particles;
+    RigidBodies m_rigidBodies;
+    ParticleContactGenerators m_particleContactGenerators;
+    ParticleForceRegistry m_particleForceRegistry;
+    ParticleGravity m_gravityForce;
+
+    Demo();
+
+    /**
+     * @brief Runs physics calculations with given duration and updates the internal states
+     * @param duration duration of the physical frame
+     */
+    void ComputeFrame(double duration);
+
+    /**
+     * @brief Renders frame and updates frame buffer
+     */
+    void RenderFrame() const;
+
+    /**
+     * @brief Makes rigid body updates internal structures and returns reference to it
+     * @param[in] particle physical data
+     * @param[in] shape a pointer to the collision geometry shape
```

Comment:
it's not really a pointer though. Also why not give the raw pointer and then just assume ownership via unique_ptr?

Reply:
I think it's a more clear sign of an intent to assume ownership, to take a unique pointer rather than a raw one.

Reply:
I just don't see any reason to create a `unique_ptr` with a sole purpose to be moved to some other method. It would make sense if this `unique_ptr` was being moved around a lot or used more than once, but in this case it is literally just create `unique_ptr` and move it right away or am I wrong?